### ISSUE TITLE
Scroll Primitives

### ISSUE THREAD

**Issue by frisitano** [scroll-tech/reth#7]
# Overview
Scroll introduces / modifies the transaction and transaction receipt primitives. We should implement these primitives in reth or a new primitives crate.

# Transaction types
Scroll supports the following transaction types:
- Type 0: Legacy Transactions
- Type 1: [EIP-2930](https://eips.ethereum.org/EIPS/eip-2930)
- Type 2: [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559)
- Type 126: L1MessageTx ([EIP-2718](https://eips.ethereum.org/EIPS/eip-2718) typed transaction)

## L1 Message Transaction (typed transaction)

L1 initiated transactions use the following transaction type associated with an EIP-2718 type of `0x7E`. There are no fees payed for this transaction type on L2 as the fees are already payed for when the transaction is submitted on L1. The encoding follows the standard EIP-2718 encoding 

```rust
struct L1MessageTX {
    queue_index: u64,
    gas: u64,
    to: Address,
    value: U256,
    data: Bytes,
    sender: Address,
}
```

# Receipt
An additional field is added to the `TransactionReceipt`. This field includes the l1 fee. See reference [here](https://github.com/scroll-tech/go-ethereum/pull/244/files#diff-4b16f7fb21c263f620aaacd67ecfed1ce1f962b105a83c9f1955f84ed1c8b984).

# Account
The account field has two additional fields `code_size` and `poseiden_code_hash`. Also the account hashing scheme is different. In the first instance I propose we add these additional fields behind a feature flag but in the future we should provide an abstraction such as providing the `Account` object in [`NodePrimites`](https://github.com/paradigmxyz/reth/blob/5e288b2469e889b41e5df16ce01951fedf4b2b55/crates/node/types/src/lib.rs#L20-L25).

# Design
Currently the [`Transaction`](https://github.com/paradigmxyz/reth/blob/c08c00f2ecea90ca9b4ca37c91a053f9eb4242ed/crates/primitives/src/transaction/mod.rs#L82-L136) type (and associated types, e.g. [`TxType`](https://github.com/paradigmxyz/reth/blob/c08c00f2ecea90ca9b4ca37c91a053f9eb4242ed/crates/primitives/src/transaction/tx_type.rs#L46-L73), [`PooledTransactionsElement`](https://github.com/paradigmxyz/reth/blob/c08c00f2ecea90ca9b4ca37c91a053f9eb4242ed/crates/primitives/src/transaction/pooled.rs#L25-L68)) and [`Receipt`](https://github.com/paradigmxyz/reth/blob/c08c00f2ecea90ca9b4ca37c91a053f9eb4242ed/crates/primitives/src/receipt.rs#L17-L46) are hardcoded. Modifications to these types are currently achieved via feature flags as seen by the `optimism` feature flag. In the first instance we should implement the scroll changes behind a feature flag but we should consult with the `reth` team to see if we can provide the necessary abstractions to allow for configurability of these types.

**Comment by Thegaram** [comment]
Edited to add "Type 1" to the supported transaction types.

**Comment by Thegaram** [comment]
Relevant reth tracking issue: https://github.com/paradigmxyz/reth/issues/11239

**Comment by greged93** [comment]
# Overview

The account structure used in Scroll contains two extra fields: `code_size` and `poseidon_code_hash`. `code_size` needs to be present for the execution and both need to be present for the computation of the state root, see following link for the [hashing scheme](https://docs.scroll.io/en/technology/sequencer/zktrie/#ethereum-account-leaf-node).

# Propositions

1. Abstract the Account. Ongoing work is being done upstream for [storage abstraction](https://github.com/paradigmxyz/reth/pull/11872) which would be required.
2. Keep the account type but compute the code size and the poseidon code hash at runtime. This might cause some overhead for the poseidon hashing of the bytecode. All this can be hidden behind the `StateRoot` GAT of the `StateCommitment`.
3. Add a feature `cfg(feature = "scroll")` for now, modifying the `Account` from Reth. This is the easiest option imo. This also currently limits the amount of change we need to get in the code, and helps us get a better feeling about how to better abstract it.

# Details

1. Account Abstraction:
- After execution of the payload, the [HashedPostState](https://github.com/paradigmxyz/reth/blob/db1d64b1c8126abf7573ccb4a0a1a757c1a100b2/crates/trie/trie/src/state.rs#L17) is populated with the accounts on which changes have occurred. These accounts would need to be abstracted and implement a trait which would allow defining the hashing of the account’s address and the encoding of the account.
- Providers modifications:
    - Modifying the [AccountReader](https://github.com/paradigmxyz/reth/blob/10f3320782354c01600347851f520de44cb2a451/crates/storage/storage-api/src/account.rs#L13) and the [StateChangeWriter](https://github.com/paradigmxyz/reth/blob/180f10001c3de93a41c690156aba28e6f6a063da/crates/storage/provider/src/traits/state.rs#L23), while adding a GAT of `Account` would allow to read and write the abstracted account from/to the database.
    - Use a `FullAccountFetcher` trait in order to populate an abstracted account right before hash computation. This fetcher could be defined to retrieve missing information from the database (Poseidon code hash, code size,…) and limit the amount of additional abstraction we need to add/propagate in the code.
2. Runtime addition: this runtime addition can be implemented in the `HashedPostState` and the `StateRoot` themselves. Changes more limited compare to the abstraction solution.
3. Feature gating: modify the type itself and define the way to convert a `BundleAccount` from revm into an `Account` with the additional fields. Changes seem limited as well.

For solution 2 and 3, needs to be verified if we can have a way to retrieve the Poseidon code hash from the `AccountInfo` structure from revm.

**Comment by frisitano** [comment]
Below I will specify the changes that could be implemented in `reth` and `revm` to support the different account structure of scroll. 

### Reth Account Abstraction
First we will introduce an `Account` trait with basic functionality and implement it on our concrete `ScrollAccount` type.
```rust
pub trait AccountT {
    const fn has_bytecode(&self) -> bool;
    fn is_empty(&self) -> bool;
    fn get_bytecode_hash(&self) -> B256;
    fn get_nonce(&self) -> u64;
    fn get_balance(&self) -> U256;
}

pub struct ScrollAccount {
    nonce: u64,
    balance: U256,
    bytecode_hash: Option<B256>,
    poseidon_bytecode_hash: Option<B256>,
    code_size: Option<usize>,
}

impl AccountT for ScrollAccount {
    const fn has_bytecode(&self) -> bool {
        self.bytecode_hash.is_some()
    }

    fn is_empty(&self) -> bool {
        self.nonce == 0 &&
            self.balance.is_zero() &&
            self.bytecode_hash.map_or(true, |hash| hash == KECCAK_EMPTY)
    }

    fn get_bytecode_hash(&self) -> B256 {
        self.bytecode_hash.unwrap_or(KECCAK_EMPTY)
    }

    fn get_nonce(&self) -> u64 {
        self.nonce
    }

    fn get_balance(&self) -> U256 {
        self.balance
    }
}
``` 

We would then provide the concrete `ScrollAccount` type in scroll `NodePrimitives` as follows:
```rust
pub trait NodePrimitives {
    type Account: AccountT;
}

impl NodePrimitives for ScrollPrimitives {
    type Account = ScrollAccount;
}
```

### Revm `AccountInfo` abstraction

In `revm` we would need to introduce a trait for `AccountInfo`. This would look something like this:

```rust
pub trait AccountInfoT {
    fn new(balance: U256, nonce: u64, code_hash: B256, code: Bytecode) -> Self;
    fn balance(&self) -> U256;
    fn nonce(&self) -> u64;
    fn code(&self) -> Option<&Bytecode>;
    fn code_hash(&self) -> B256;
    fn set_code(&mut self, code: Bytecode);
    fn set_balance(&mut self, balance: U256);
    fn set_nonce(&mut self, nonce: u64);
    fn copy_without_code(&self) -> Self;
    fn is_empty(&self) -> bool;
    fn exists(&self) -> bool;
    fn has_no_code_and_nonce(&self) -> bool;
    fn is_empty_code_hash(&self) -> bool;
    fn take_bytecode(&mut self) -> Option<Bytecode>;
    fn from_balance(balance: U256) -> Self;
    fn from_bytecode(bytecode: Bytecode) -> Self;
}

pub struct ScrollAccountInfo {
    balance: U256,
    nonce: u64,
    code_hash: B256,
    code: Option<Bytecode>,
    poseidon_code_hash: B256,
    code_size: usize,
}

impl AccountInfoT for ScrollAccountInfo {
    ...
}
```

We would then provide the concrete `Account` type by adding a new associated type in `EvmWirign` and specifying the `ScrollAccountInfo` type when configuring the evm.

```rust
pub trait EvmWiring: Sized {
    /// External context type
    type ExternalContext: Sized;

    /// Chain context type.
    type ChainContext: Sized + Default + Debug;

    /// Database type.
    type Database: Database;

    /// Account type.
    type Account: AccountT;

    /// The type that contains all block information.
    type Block: Block;

    /// The type that contains all transaction information.
    type Transaction: Transaction;

    /// The type that enumerates the chain's hardforks.
    type Hardfork: HardforkTrait;

    /// Halt reason type.
    type HaltReason: HaltReasonTrait;
}
```

This is a very invasive change and impacts many different data structures within `revm` so I think this is a no-go. 

### `ScrollStateProviderDatabase` and `BundleState` context
This solution would mitigate any changes needed in `revm` and would keep the additional account fields encapsulated in `reth`. 

The idea would be to introduce a new `ScrollStateProviderDatabase` which is a modified version of [`StateProviderDatabase`](https://github.com/paradigmxyz/reth/blob/e3173407e103fd036ab6240d599668292cf1190a/crates/revm/src/database.rs#L68). This stateful object would include a cache for scroll code fields which is `code_hash -> (code_size, poseidon_code_size)` as seen below:

```rust
pub struct ScrollStateProviderDatabase<DB>{
    scroll_code: HashMap<B256, (usize, B256)>,   
    pub db: DB
}
```

For the implementation of `DatabaseRef` on `ScrollStateProviderDatabase` we would cache the scroll code fields before returning the `AccountInfo` as follows:

```rust
impl<DB: EvmStateProvider> DatabaseRef for ScrollStateProviderDatabase<DB> {
    type Error = <Self as Database>::Error;

    /// Retrieves basic account information for a given address.
    ///
    /// Returns `Ok` with `Some(AccountInfo)` if the account exists,
    /// `None` if it doesn't, or an error if encountered.
    fn basic_ref(&self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
        let account = self.basic_account(address)?;
        let (account_info, scroll_code) = account.into_parts();
        self.scroll_code.insert(account.info.code_hash, scroll_code);
        Ok(account_info)
    }
```

We then look to introduce an additional generic `context` field on `BundleState`, hopefully upstream should support this as the surface area shouldn't be too large.

```rust
pub struct BundleState<C> {
    /// Account state.
    pub state: HashMap<Address, BundleAccount>,
    /// All created contracts in this block.
    pub contracts: HashMap<B256, Bytecode>,
    /// Additional context associated with the bundle state.
    pub context: C,
    /// Changes to revert.
    ///
    /// Note: Inside vector is *not* sorted by address.
    /// But it is unique by address.
    pub reverts: Reverts,
    /// The size of the plain state in the bundle state.
    pub state_size: usize,
    /// The size of reverts in the bundle state.
    pub reverts_size: usize,
}
```

Then on [`State`](https://github.com/bluealloy/revm/blob/bbc8d81dbe2a6a4d184e32fa540e1c4e248a65c9/crates/database/src/states/state.rs#L23-L58) we can implement a trait `TakeScrollCode` which will take the scroll code map from the database and store it in `BundleState`. 

```rust
pub trait MergeScrollCode {
    fn merge_scroll_code(&mut self, code_hash: B256);
}

impl<DB> MergeScrollCode for State<DB> {
    fn merge_scroll_code(&mut self, code_hash: B256) {
        let scroll_code = self.db.scroll_code.take();
        self.bundle_state.context = scroll_code;
    }
}
```

This will mean that all code size and poseiden code hashes that were read from the database are available in `BundleState`. For newly deployed contract code we will have the bytecode available in the `BundleState` such that we can compute the poseiden hash and code size at runtime. Therefore when we convert from `BundleState` into `HashedPostState` which contains a mapping of `ScrollAccount`'s we will have all of the code information available in the environment to populate those additional fields.

**Comment by greged93** [comment]
I agree with the second solution. Just two nits:

1. I think we need `Database` and not `DatabaseRef` implemented, here is the trait bound on the `BlockExecutorProvider`:
```rust
    /// Creates a new executor for single block execution.
    ///
    /// This is used to execute a single block and get the changed state.
    fn executor<DB>(&self, db: DB) -> Self::Executor<DB>
    where
        DB: Database<Error: Into<ProviderError> + Display>;

```
2. While waiting for the update to the `BundleState` to be merged upstream, I propose introducing our own `ScrollBundleState`:
```rust
pub struct ScrollBundleState {
    context: HashMap<B256, (usize, B256)>,
    bundle: BundleState
}
```
which we can re-export via the [`reth_evm`](https://github.com/paradigmxyz/reth/blob/5af551782c9084bcd61d15529b167b90bcc86c9c/crates/revm/src/lib.rs#L28) crate:
```rust
// Convenience re-exports.
pub use revm::{self, *};
#[cfg(feature = "scroll")]
pub use revm_scroll::ScrollBundleState as BundleState
```

**Comment by frisitano** [comment]
> I agree with the second solution. Just two nits:
> 
> 1. I think we need `Database` and not `DatabaseRef` implemented, required by `BlockExecutorProvider` for [live sync](https://github.com/paradigmxyz/reth/blob/3f4634ccbc4cbf64720627872ead3155259ce9cc/crates/engine/tree/src/tree/mod.rs#L2180) block execution.

Agreed, the `DatabaseRef` trait wouldn't work as it takes an immutable `&self` reference but we need to modify the code cache. 

> 2. While waiting for the update to the `BundleState` to be merged upstream, I propose introducing our own `ScrollBundleState`:
> 
> ```rust
> pub struct ScrollBundleState {
>     context: HashMap<B256, (usize, B256)>,
>     bundle: BundleState
> }
> ```

Looks reasonable to me. The only question I have is do we want to unpack fields from `BundleState` such that they are at the root level of `ScrollBundleState`. I haven't looked at the code but I wonder if there are places in the code where public fields of `BundleState` are accessed directly.

I would propose that we open an issue in upstream `revm` to suggest the introduction of this new generic context field to get feedback. Would you be able to do that please @greged93?

**Comment by greged93** [comment]
After some research on the areas of the codebase affected by the above changes, I came across the below related to persistence:

- During both live sync and pipelining, state is persisted into the database using the [StateChangeWriter](https://github.com/paradigmxyz/reth/blob/c6b740801fda7963baeb33ac395227032825bb7e/crates/storage/provider/src/providers/database/provider.rs#L2176).
- Methods `write_state_reverts` and `write_state_changes` from the trait receive `PlainStateReverts` and `StateChangeset` respectively. These structures don't have access to [Account](https://github.com/paradigmxyz/reth/blob/5568cca846d01dce490f0b65656d323689d9d4fe/crates/primitives-traits/src/account.rs#L30-L37) but rather [AccountInfo](https://github.com/bluealloy/revm/blob/release/v46/crates/primitives/src/state.rs#L16-L23) from revm.
- Method `write_hashed_state` receives a [HashedPostStateSorted](https://github.com/paradigmxyz/reth/blob/db1d64b1c8126abf7573ccb4a0a1a757c1a100b2/crates/trie/trie/src/state.rs#L274-L279) which does use the [Account](https://github.com/paradigmxyz/reth/blob/5568cca846d01dce490f0b65656d323689d9d4fe/crates/primitives-traits/src/account.rs#L30-L37).
- This means we wouldn't currently be able to maintain correct state in database for the `code_size` and the `poseidon_code_hash` fields when writing reverts or changes, while it is possible for the hashed state.

The below details a proposed approach to abstract the `BundleState` completely, in order to avoid issues with the `StateChangeWriter`:

Add a `ExecutionPrimitives` GAT into the `NodePrimitives`:
```rust
pub trait NodePrimitives {
    /// Block primitive.
    type Block;
    /// Node execution primitives
    type ExecutionPrimitives: NodeExecutionPrimitives;
}
```
We then bound the `ExecutionPrimitives` GAT with `NodeExecutionPrimitives`, which we use to define an additional execution context and a `BundleState`.
```rust
pub trait NodeExecutionPrimitives {
    /// Additional context extracted during execution
    type Context;
    /// Bundle state of values changed during execution
    type BundleState: From<(revm::db::BundleState, NodeExecutionPrimitives::Context)> + BundleStateT;
}

pub trait BundleStateT {
    type Reverts;
    type StateChangeset;
    ...
}
```

As an example, for Scroll, we would implement the following:
```rust
type KeccakHash = B256;
type PoseidonHash = B256;
type CodeSize = usize;
type ScrollExecutionContext = HashMap<KeccakHash, (CodeSize, PoseidonHash)>;

pub struct ScrollBundleState {
    context: ScrollExecutionContext,
    ...,
}

impl From<(revm::db::BundleState, ScrollExecutionContext)> for ScrollBundleState {
    fn from(value: (BundleState, ScrollExecutionContext)) -> Self {
        Self {
            context: value.1,
            ...,
        }
    }
}

pub struct ScrollExecutionPrimitives;

impl NodeExecutionPrimitives for ScrollExecutionPrimitives {
    type Context = ScrollExecutionContext;
    type BundleState = ScrollBundleState;
}
```

Using the `ScrollStateProviderDatabase`, we can then retrieve this context and convert the `revm::BundleState` to the `NodeExecutionPrimitives::BundleState`.

We will start implementing the above when time allows it and pushing it to upstream, but in the meantime will progress by defining our own `ScrollBundleState` and re-exporting it via a feature flag in `reth_revm`. 
The `ScrollBundleState` would use a `ScrollAccountInfo` in place of the `AccountInfo` from `revm`, as well as our own `ScrollPlainStateReverts` and `ScrollStateChangeset`, which allows us to avoid the afore mentioned issue.

**Comment by frisitano** [comment]
This looks good to me, just a few points. Do we need to introduce `NodeExecutionPrimitives` in `NodePrimitives` or can the concrete `BundleState` type be inferred by what is returned from the `Executor`? Maybe we just need to add the `BundleState` generic on `ExecutionOutcome`?

For the `ScrollBundleState` do we need to store the `context` field or can we just parse it during the conversion `From<(revm::db::BundleState, ScrollExecutionContext)> for ScrollBundleState` such that we construct `ScrollAccountInfo` in the `BundleAccount`'s within `ScrollBundleState`?

**Comment by greged93** [comment]
> This looks good to me, just a few points. Do we need to introduce NodeExecutionPrimitives in NodePrimitives or can the concrete BundleState type be inferred by what is returned from the Executor? Maybe we just need to add the BundleState generic on ExecutionOutcome?

After some discussion, the following route has been chosen:

Take advantage of the `Output` GAT on `Executor` and `BatchExecutor` by extracting them one level higher in the `BlockExecutorProvider`:

```rust
pub trait BlockExecutorProvider: Send + Sync + Clone + Unpin + 'static {
    type ExecutorOutput;
    type BatchExecutorOutput;

    type Executor<DB: Database<Error: Into<ProviderError> + Display>>: for<'a> Executor<
        DB,
        Input<'a> = BlockExecutionInput<'a, BlockWithSenders>,
        Output = Self::ExecutorOutput,
        Error = BlockExecutionError,
    >;
    type BatchExecutor<DB: Database<Error: Into<ProviderError> + Display>>: for<'a> BatchExecutor<
        DB,
        Input<'a> = BlockExecutionInput<'a, BlockWithSenders>,
        Output = BatchExecutorOutput,
        Error = BlockExecutionError,
    >;
    ...
```

This removes the need for the `NodeExecutionPrimitives` and allows us to define in our `BlockExecutorProvider` the structure returned after execution. 

We will need to bound the `ExecutorOutput` and the `BatchExecutorOutput` by an additional trait `ExecutionOutcome` which would provide some fetching functionality. A fitting candidate would be the [`ExecutionDataProvider`](https://github.com/paradigmxyz/reth/blob/0c70f6bd3519d5c5bce85481bea728df89fdabf3/crates/storage/storage-api/src/state.rs#L203-L208), which would need to be slightly updated:
```rust
pub trait ExecutionDataProvider {
    type Reverts;
    type StateChangeset;

    /// Return the execution outcome.
    fn execution_outcome(&self) -> &ExecutionOutcome;
    /// Return block hash by block number of pending or canonical chain.
    fn block_hash(&self, block_number: BlockNumber) -> Option<BlockHash>;
}
```

**Comment by greged93** [comment]
> This looks good to me, just a few points. Do we need to introduce NodeExecutionPrimitives in NodePrimitives or can the concrete BundleState type be inferred by what is returned from the Executor? Maybe we just need to add the BundleState generic on ExecutionOutcome?

After some discussion, the following route has been chosen:

Take advantage of the `Output` GAT on `Executor` and `BatchExecutor` by extracting them one level higher in the `BlockExecutorProvider`:

```rust
pub trait BlockExecutorProvider: Send + Sync + Clone + Unpin + 'static {
    type ExecutorOutput;
    type BatchExecutorOutput;

    type Executor<DB: Database<Error: Into<ProviderError> + Display>>: for<'a> Executor<
        DB,
        Input<'a> = BlockExecutionInput<'a, BlockWithSenders>,
        Output = Self::ExecutorOutput,
        Error = BlockExecutionError,
    >;
    type BatchExecutor<DB: Database<Error: Into<ProviderError> + Display>>: for<'a> BatchExecutor<
        DB,
        Input<'a> = BlockExecutionInput<'a, BlockWithSenders>,
        Output = BatchExecutorOutput,
        Error = BlockExecutionError,
    >;
    ...
```

This removes the need for the `NodeExecutionPrimitives` and allows us to define in our `BlockExecutorProvider` the structure returned after execution. 

We will need to bound the `ExecutorOutput` and the `BatchExecutorOutput` by an additional trait `ExecutionOutcome` which would provide some fetching functionality. A fitting candidate would be the [`ExecutionDataProvider`](https://github.com/paradigmxyz/reth/blob/0c70f6bd3519d5c5bce85481bea728df89fdabf3/crates/storage/storage-api/src/state.rs#L203-L208), which would need to be slightly updated:
```rust
pub trait ExecutionDataProvider {
    type StateChanges;
    type Reverts;
    type PlainStateChanges;

    /// Return the reverts from execution.
    fn reverts(&self) -> &Self::Reverts;
    /// Returns the complete state changes from execution.
    fn state_changes(&self) -> &Self::StateChanges;
    /// Returns a reduced view of changes for database commitment.
   fn plain_state_changes(&self) -> &Self::PlainStateChanges;
}
```

**Comment by greged93** [comment]
@frisitano I think we can close this

**Comment by frisitano** [comment]
> @frisitano I think we can close this

agreed.

closed by #39, #33, #41, #42
